File: python/packages/spore_core/src/spore_core/tasklist.py
"""Issue #71 — Persisted task-list tool (PlanExecute, drives the execute loop).

Mirrors the Rust reference at ``rust/crates/spore-core/src/tasklist.rs``.

Decomposed out of #59. The accepted plan (#70) is parsed into a persisted task
list (#72), and the execute phase (#59) loops over the tasks until the list is
complete. This module owns the **task-list primitive** that holds and mutates
that list, plus its disk-persistence helpers. The single mutating tool lives in
:mod:`spore_tools.tools.tasklist`. It is consumed by #72 (which populates the
list) and #59 (whose execute loop drains it).

Types
-----
* :class:`TaskStatus` — ``pending | in_progress | completed | blocked``.
  Serializes to those exact snake_case strings. Byte-identical across all four
  languages.
* :class:`Task` — ``{id: int, description: str, status: TaskStatus,
  blockers: list[int]}``. A flat record: NO hierarchy/subtasks, NO timestamps
  (byte-identity constraint), NO order field (order is positional in
  :attr:`TaskList.tasks`). ``blockers`` (#118) are ids of other tasks that must
  be ``completed`` before this task runs; it is the LAST wire field and defaults
  to empty so pre-#118 blobs still load. Empty blockers ALWAYS serialize as
  ``[]`` (never dropped), mirroring ``next_id``.
* :class:`TaskList` — ``{tasks: list[Task], next_id: int}``. The persisted
  collection. The default is empty with ``next_id == 1``.
* :class:`TaskListError` — domain error
  (:meth:`TaskListError.task_not_found`, :meth:`TaskListError.invalid_transition`,
  :meth:`TaskListError.invalid_blockers`). These map to a recoverable tool error
  at the tool boundary; the tool NEVER raises uncaught.
* :class:`BlockerRejection` — why a blocker set was rejected by
  :meth:`TaskList.add`: ``self_block`` | ``unknown_id`` | ``cycle``.

ID scheme
---------
Sequential ``int``, 1-based, assigned monotonically from
:attr:`TaskList.next_id`. Ids are never reused — ``next_id`` only ever grows,
and it is preserved across reload so a freshly-loaded list keeps minting fresh
ids.

Rules enforced
--------------
* R1  Ids are assigned 1, 2, 3, … monotonically from ``next_id``; never reused.
* R2  :meth:`TaskList.add` APPENDS to the end of ``tasks`` (positional order is
  stable).
* R3  Listing never mutates the list.
* R4  Unknown id on update/complete → :meth:`TaskListError.task_not_found`
  (recoverable at the tool boundary).
* R5  Status transitions follow the matrix in :func:`validate_transition`
  (DECISION 1, "permissive-except-terminal-Completed"). A rejected transition →
  :meth:`TaskListError.invalid_transition`.
* R6  ``completed`` is terminal: ANY transition OUT of ``completed`` is rejected
  (the idempotent ``completed → completed`` self-transition is allowed).
* R7  Self-transitions ``X → X`` are idempotent and always allowed.
* R9  (#118) :meth:`TaskList.add` blockers are validated BEFORE any mutation; a
  reject leaves the list untouched (mirrors :meth:`TaskList.update`). Validation:
  self-block (a blocker == the about-to-be-assigned id) →
  :meth:`TaskListError.invalid_blockers` (``self_block``); unknown id (a blocker
  matching no existing task) → (``unknown_id`` with the offending ``blocker``);
  cycle (the new edges would close a directed cycle in the blockers graph,
  checked by :func:`would_create_cycle`) → (``cycle``). Empty blockers never
  reject.
* R8  Persistence is through the storage seam (#75): the standalone
  :class:`~spore_tools.tools.TaskListTool` persists via the
  :class:`~spore_core.storage.RunStore` on the ``ToolContext``, keyed by
  :class:`SessionId` under :data:`TASK_LIST_EXTRAS_KEY`. The retired interim
  sandbox path (``.spore/task_list.json``) is GONE; with the library's default
  no-op storage a standalone tool call persists nothing across processes (an
  accepted behavior change — no migration shim). #59's execute loop shares the
  same ``RunStore`` key.

Both design forks (transition matrix, state seam) were resolved before
implementation.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import Enum

from .errors import SporeError
from .hooks import PlanArtifact

__all__ = [
    "TASK_LIST_EXTRAS_KEY",
    "BlockerRejection",
    "Task",
    "TaskList",
    "TaskListError",
    "TaskStatus",
    "plan_artifact_to_task_list",
    "validate_transition",
    "would_create_cycle",
]

#: Key under which the :class:`TaskList` is persisted in the
#: :class:`~spore_core.storage.RunStore`, keyed by :class:`SessionId`. Both the
#: harness-side #59 execute loop and the standalone
#: :class:`~spore_tools.tools.TaskListTool` (#75) share this single key, so a
#: standalone tool call and a PlanExecute run on the same session intentionally
#: share one blob. Stable across all four languages. The JSON shape is the
#: canonical serialized :class:`TaskList` (``{"tasks":[...],"next_id":N}``).
TASK_LIST_EXTRAS_KEY = "task_list"


# ============================================================================
# Types
# ============================================================================


class TaskStatus(str, Enum):
    """Lifecycle status of a :class:`Task`. Serializes to its snake_case value."""

    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    #: Waiting on a blocker. (#118) This means BOTH "waiting on a blocker that
    #: has not yet completed" AND "a blocker failed terminally" — the status is
    #: the same in either case; the distinction (if any) lives in the scheduler,
    #: not the schema.
    BLOCKED = "blocked"


@dataclass
class Task:
    """A single task: flat, no hierarchy, no timestamps, no order field (order is
    positional in :attr:`TaskList.tasks`).

    ``blockers`` (#118) are the ids of other tasks that must be
    :attr:`TaskStatus.COMPLETED` before this task runs. The canonical wire field
    order is ``id, description, status, blockers`` (blockers LAST). ``blockers``
    defaults to empty so a pre-#118 blob without the key still deserializes (to
    an empty list); empty blockers ALWAYS serialize as ``[]`` (never dropped),
    the same treatment as :attr:`TaskList.next_id`.
    """

    id: int
    description: str
    status: TaskStatus
    blockers: list[int] = field(default_factory=list)

    def to_dict(self) -> dict[str, object]:
        """Serialize to a plain dict with canonical field order ``id``,
        ``description``, ``status``, ``blockers`` (matching the cross-language
        wire form). Empty ``blockers`` always serialize as ``[]``."""
        return {
            "id": self.id,
            "description": self.description,
            "status": self.status.value,
            "blockers": list(self.blockers),
        }

    @staticmethod
    def from_dict(value: dict[str, object]) -> Task:
        # ``blockers`` is optional on the wire (pre-#118 blob) → default empty.
        blockers_value = value.get("blockers", [])
        if not isinstance(blockers_value, list):
            raise ValueError("field `blockers` is not an array")
        return Task(
            id=int(value["id"]),  # type: ignore[arg-type]
            description=str(value["description"]),
            status=TaskStatus(value["status"]),
            blockers=[int(b) for b in blockers_value],
        )


@dataclass
class TaskList:
    """The persisted collection of tasks plus the monotonic id counter.

    Serializes as ``{"tasks":[...],"next_id":N}``. The default (and every
    freshly-minted list) starts at ``next_id == 1``. A deserialized blob without
    ``next_id`` defaults to ``0`` (older/handwritten form), but the constructed
    default starts at ``1``.
    """

    tasks: list[Task] = field(default_factory=list)
    next_id: int = 1

    # ---- serialization ----------------------------------------------------

    def to_dict(self) -> dict[str, object]:
        """Canonical dict form: ``tasks`` then ``next_id`` (field order is wire
        significant for byte-identity)."""
        return {
            "tasks": [t.to_dict() for t in self.tasks],
            "next_id": self.next_id,
        }

    def to_json(self) -> str:
        """Compact, byte-identical JSON: separators ``(",", ":")`` and the field
        order pinned by :meth:`to_dict`. Matches the Rust ``serde_json`` form."""
        return json.dumps(self.to_dict(), separators=(",", ":"))

    @staticmethod
    def from_dict(value: dict[str, object]) -> TaskList:
        tasks_value = value.get("tasks", [])
        if not isinstance(tasks_value, list):
            raise ValueError("field `tasks` is not an array")
        tasks = [Task.from_dict(t) for t in tasks_value]
        # ``next_id`` is optional on the wire (older blob) → default 0.
        next_id = int(value.get("next_id", 0))  # type: ignore[arg-type]
        return TaskList(tasks=tasks, next_id=next_id)

    @staticmethod
    def from_json(text: str) -> TaskList:
        value = json.loads(text)
        if not isinstance(value, dict):
            raise ValueError("top-level JSON value is not an object")
        return TaskList.from_dict(value)

    # ---- mutation helpers (the seam #72 / #59 will call) ------------------

    def add(self, description: str, blockers: list[int] | None = None) -> int:
        """Append a new ``pending`` task with the next sequential id and return
        that id. Increments :attr:`next_id`. R1, R2.

        Fallible since #118: ``blockers`` are validated BEFORE any mutation, so a
        rejected blocker set leaves the list completely untouched (mirroring how
        :meth:`update` validates before writing). R9. Validation order:

        1. self-block — a blocker equal to the id about to be assigned
           (``next_id``) → :meth:`TaskListError.invalid_blockers` (``self_block``).
        2. unknown id — a blocker matching no existing task id → (``unknown_id``).
        3. cycle — the new edges would close a directed cycle, checked by
           :func:`would_create_cycle` → (``cycle``).

        Empty ``blockers`` always pass (and serialize as ``[]``).
        """
        blocker_ids = list(blockers) if blockers is not None else []
        task_id = self.next_id

        for blocker in blocker_ids:
            if blocker == task_id:
                raise TaskListError.invalid_blockers(task_id, BlockerRejection.self_block())
            if not any(t.id == blocker for t in self.tasks):
                raise TaskListError.invalid_blockers(task_id, BlockerRejection.unknown_id(blocker))

        if would_create_cycle(self.tasks, task_id, blocker_ids):
            raise TaskListError.invalid_blockers(task_id, BlockerRejection.cycle())

        self.tasks.append(
            Task(
                id=task_id,
                description=description,
                status=TaskStatus.PENDING,
                blockers=blocker_ids,
            )
        )
        self.next_id += 1
        return task_id

    def update(
        self,
        task_id: int,
        status: TaskStatus | None = None,
        description: str | None = None,
    ) -> None:
        """Update a task's status and/or description.

        * Unknown id → raises :class:`TaskListError` (``task_not_found``).
        * ``status`` present → validated via :func:`validate_transition` then
          applied.
        * ``description`` present → set verbatim.
        * Both absent → no-op success.

        Status is validated BEFORE any field is written, so a rejected
        transition leaves the task untouched.
        """
        task = self._find(task_id)
        if status is not None:
            validate_transition(task_id, task.status, status)
            task.status = status
        if description is not None:
            task.description = description

    def complete(self, task_id: int) -> None:
        """Mark a task ``completed``, validating the transition first.

        * Unknown id → raises :class:`TaskListError` (``task_not_found``).
        * Already ``completed`` → idempotent success.
        """
        task = self._find(task_id)
        validate_transition(task_id, task.status, TaskStatus.COMPLETED)
        task.status = TaskStatus.COMPLETED

    def _find(self, task_id: int) -> Task:
        for task in self.tasks:
            if task.id == task_id:
                return task
        raise TaskListError.task_not_found(task_id)


# ============================================================================
# Errors
# ============================================================================


@dataclass(frozen=True)
class BlockerRejection:
    """Why an ``add_task`` blockers set was rejected (#118).

    A single value type with a ``reason`` discriminant
    (``"self_block"`` | ``"unknown_id"`` | ``"cycle"``), mirroring the Rust
    ``BlockerRejection`` enum (internally tagged on ``reason``). ``blocker`` is
    populated only for ``unknown_id`` (the offending id).
    """

    #: One of ``"self_block"`` | ``"unknown_id"`` | ``"cycle"``.
    reason: str
    #: The offending blocker id (``unknown_id`` only); ``None`` otherwise.
    blocker: int | None = None

    @staticmethod
    def self_block() -> BlockerRejection:
        """A blocker referenced the id about to be assigned to this very task."""
        return BlockerRejection(reason="self_block")

    @staticmethod
    def unknown_id(blocker: int) -> BlockerRejection:
        """A blocker referenced an id that matches no existing task."""
        return BlockerRejection(reason="unknown_id", blocker=blocker)

    @staticmethod
    def cycle() -> BlockerRejection:
        """The new blocker edges would close a directed cycle in the graph."""
        return BlockerRejection(reason="cycle")

    def to_dict(self) -> dict[str, object]:
        """Canonical dict form, internally tagged on ``reason``. ``unknown_id``
        also carries the offending ``blocker``."""
        if self.reason == "unknown_id":
            return {"reason": "unknown_id", "blocker": self.blocker}
        return {"reason": self.reason}

    def __str__(self) -> str:
        if self.reason == "self_block":
            return "a task cannot block itself"
        if self.reason == "unknown_id":
            return f"unknown blocker id: {self.blocker}"
        return "blocker edges would create a cycle"


class TaskListError(SporeError):
    """Errors raised by task-list mutations. Every variant is recoverable at the
    tool boundary.

    A single class with a ``kind`` discriminant
    (``"task_not_found"`` | ``"invalid_transition"`` | ``"invalid_blockers"``),
    mirroring the Rust ``TaskListError`` enum.
    """

    def __init__(
        self,
        message: str,
        *,
        kind: str,
        id: int,
        from_status: TaskStatus | None = None,
        to_status: TaskStatus | None = None,
        reason: BlockerRejection | None = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.kind = kind
        self.id = id
        self.from_status = from_status
        self.to_status = to_status
        self.reason = reason

    @staticmethod
    def task_not_found(task_id: int) -> TaskListError:
        return TaskListError(f"task not found: {task_id}", kind="task_not_found", id=task_id)

    @staticmethod
    def invalid_transition(
        task_id: int, from_status: TaskStatus, to_status: TaskStatus
    ) -> TaskListError:
        return TaskListError(
            f"invalid transition for task {task_id}: {from_status.value} -> {to_status.value}",
            kind="invalid_transition",
            id=task_id,
            from_status=from_status,
            to_status=to_status,
        )

    @staticmethod
    def invalid_blockers(task_id: int, reason: BlockerRejection) -> TaskListError:
        return TaskListError(
            f"invalid blockers for task {task_id}: {reason}",
            kind="invalid_blockers",
            id=task_id,
            reason=reason,
        )


# ============================================================================
# Transition matrix (DECISION 1)
# ============================================================================

# Allowed non-self transitions under DECISION 1
# ("permissive-except-terminal-Completed"). Self-transitions ``X → X`` are
# handled separately (always allowed). ANY transition OUT of ``completed`` is
# rejected (it is terminal).
_ALLOWED_TRANSITIONS: frozenset[tuple[TaskStatus, TaskStatus]] = frozenset(
    {
        (TaskStatus.PENDING, TaskStatus.IN_PROGRESS),
        (TaskStatus.PENDING, TaskStatus.COMPLETED),
        (TaskStatus.PENDING, TaskStatus.BLOCKED),
        (TaskStatus.IN_PROGRESS, TaskStatus.COMPLETED),
        (TaskStatus.IN_PROGRESS, TaskStatus.BLOCKED),
        (TaskStatus.BLOCKED, TaskStatus.IN_PROGRESS),
        (TaskStatus.BLOCKED, TaskStatus.COMPLETED),
    }
)


def validate_transition(task_id: int, from_status: TaskStatus, to_status: TaskStatus) -> None:
    """Validate a status transition under DECISION 1.

    Allowed:

    * any self-transition ``X → X`` (idempotent),
    * ``pending → in_progress | completed | blocked``,
    * ``in_progress → completed | blocked``,
    * ``blocked → in_progress | completed``.

    Rejected: ANY transition OUT of ``completed`` (it is terminal) — except the
    idempotent ``completed → completed``.

    Raises :class:`TaskListError` (``invalid_transition``) on a rejected
    transition. ``task_id`` only populates the error.
    """
    # Idempotent self-transition always allowed (incl. completed -> completed).
    if from_status == to_status:
        return
    if (from_status, to_status) in _ALLOWED_TRANSITIONS:
        return
    raise TaskListError.invalid_transition(task_id, from_status, to_status)


# ============================================================================
# Cycle detection (#118)
# ============================================================================


def would_create_cycle(tasks: list[Task], new_id: int, new_blockers: list[int]) -> bool:
    """Would adding a node ``new_id`` whose outgoing blocker edges are
    ``new_blockers`` close a directed cycle in the blockers graph of ``tasks``?

    The graph is ``task -> blocker`` (a task points at each id it is blocked by).
    A cycle exists if, starting from any of the new edges' targets, a directed
    path leads back to ``new_id``. Since a single append-only :meth:`TaskList.add`
    only references EARLIER real ids, this can never actually fire in normal use;
    the helper exists as a spec acceptance criterion (#118) and is unit-tested
    directly against a hand-built cyclic graph.
    """
    stack: list[int] = list(new_blockers)
    visited: set[int] = set()
    by_id = {t.id: t for t in tasks}
    while stack:
        node = stack.pop()
        if node == new_id:
            return True
        if node in visited:
            continue
        visited.add(node)
        task = by_id.get(node)
        if task is not None:
            stack.extend(task.blockers)
    return False


# ============================================================================
# Plan → TaskList parser (issue #72; the bridge between #70 and #59)
# ============================================================================


def plan_artifact_to_task_list(artifact: PlanArtifact) -> TaskList:
    """Parse an accepted :class:`PlanArtifact` (#70) into a fresh, ready-to-persist
    :class:`TaskList` (#71).

    This is the bridge between the plan phase and the execute loop: once a plan
    is produced and accepted, its steps become the task list that #59's execute
    loop drains.

    Types bridged
    -------------
    * Input: :class:`PlanArtifact` ``{tasks: list[str], rationale: str}``.
    * Output: :class:`TaskList` ``{tasks: list[Task], next_id: int}``.

    Rules enforced
    --------------
    * One :class:`Task` per plan step, in plan order (positional, via
      :meth:`TaskList.add`).
    * Every produced task is :attr:`TaskStatus.PENDING`.
    * Step descriptions are copied VERBATIM — no trim, no normalize, no filter
      (even ``"  spaced  "`` and ``""`` are kept).
    * Ids are assigned ``1..=n`` sequentially via the :attr:`TaskList.next_id`
      scheme; ``next_id`` ends at ``n + 1``.
    * An empty plan (``tasks: []``) yields the empty default ``TaskList``
      (``{tasks: [], next_id: 1}``). That is a valid EMPTY list, not an error
      and not "immediate completion"; the execute loop (#59) decides loop
      semantics.
    * ``rationale`` is DROPPED — neither :class:`Task` nor :class:`TaskList`
      carries it.

    Pure and total: ``PlanArtifact -> TaskList``, no async, no I/O, never raises.
    The same artifact always yields the same task list, so the mapping is
    byte-identical across all four languages.

    Always builds a fresh default :class:`TaskList`; it never merges into an
    existing list (replanning is out of scope — single parse per accepted plan).
    Wiring this into the plan-acceptance seam is DEFERRED to #59's execute loop;
    #72 ships only this pure function.
    """
    task_list = TaskList()  # next_id == 1
    for step in artifact.tasks:
        # verbatim; appends pending; bumps next_id. Empty blockers can never
        # reject, so `add` never raises here and the parser stays total. (#118)
        task_list.add(step)
    return task_list

File: go/spore-core/tools/tasklist.go
// TaskList tool (#71, storage seam #75): the single mutating tool over the
// persisted task list.
//
// One tool, TaskListTool (NAME = "task_list"), dispatched on an `action`
// discriminator (add_task, update_task, complete_task, list_tasks). See the
// parent package's tasklist.go for the types and the transition matrix.
//
// # Storage seam (#75)
//
// The tool persists via the *ToolContext's RunStore — NOT the sandbox
// filesystem. It is read-modify-write keyed by the STABLE project namespace
// (#142 — ToolContext.ProjectNamespace, falling back to SessionID when unset)
// under TaskListExtrasKey ("task_list"):
//  1. parse params (bad input → recoverable error),
//  2. toolCtx.GetDurable(ctx, "task_list") (absent → DefaultTaskList),
//  3. apply the action (domain errors → recoverable),
//  4. on a mutating action, toolCtx.PutDurable(ctx, "task_list", value),
//  5. return the serialized current list as success content (on add_task,
//     prefixed with the assigned id as a leading `added` key — see #143 below).
//
// Keying by the project namespace (not the per-window SessionID) is the #142 fix:
// the Ralph wrapper mints a fresh SessionID each context window, so a
// session-keyed list was orphaned at every window boundary; a project-keyed list
// survives window resets AND process restarts (with a durable RunStore backend).
//
// # add_task surfaces the assigned id (#143)
//
// On a successful add_task, the success content is the canonical TaskList object
// with ONE extra top-level key `added` placed FIRST — the id just assigned by
// TaskList.Add:
//
//	{"added":3,"tasks":[...],"next_id":4}
//
// The field order is EXACTLY `added`, then `tasks`, then `next_id`, and is
// byte-identical across all four languages so a model can reference a just-added
// task without re-parsing the whole list or predicting ids. The `added` key
// appears ONLY on the add_task success branch — update_task, complete_task, and
// list_tasks keep returning the bare serialized TaskList ({"tasks":[...],
// "next_id":N}), unchanged. A rejected add_task (self-block / unknown blocker /
// cycle) still returns a recoverable error with NO `added` and no list.
//
// CRITICAL: the PERSISTED RunStore blob stays EXACTLY {"tasks":[...],"next_id":N}
// — NO `added` key. `added` lives only in the tool's success content, never in
// what is persisted; the PlanExecute executor depends on the persisted blob shape.
//
// Shared key: this standalone tool and the harness-side PlanExecute execute loop
// (#76) persist under the SAME RunStore key ("task_list"), keyed by the project
// namespace (#142). A standalone tool call and a PlanExecute run on the same
// project intentionally share one blob. The JSON shape is the canonical TaskList
// serialization ({"tasks":[...],"next_id":N}), unchanged.
//
// Behavior change vs the retired sandbox path: previously the tool persisted to
// .spore/task_list.json via the sandbox. That path is GONE. With the library's
// default storage (storage.NoOp / a nil RunStore on the ToolContext) a standalone
// tool call persists NOTHING across processes — the no-op run store silently
// discards writes and returns "not found" on read. This is an accepted behavior
// change: durable cross-process persistence now requires configuring a real
// StorageProvider (and injecting it via StandardToolRegistry.SetToolContext).
// There is NO migration shim for old on-disk .spore/task_list.json files.
//
// Storage-error mapping: a storage Get/Put error maps to a recoverable
// ToolOutput error, as does a present-but-malformed blob (parse failure).
// list_tasks never writes.
//
// CRITICAL: this tool is NOT annotated ReadOnly. Read-only tools are run
// CONCURRENTLY by DispatchAll, and a concurrent read-modify-write over the same
// key would race. Leaving ReadOnly false makes the registry dispatch it
// sequentially. Destructive / OpenWorld are also left false so it is not
// treated as an irreversible side effect.

package tools

import (
	"context"
	"encoding/json"
	"fmt"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// TaskListToolName is the registered tool name.
const TaskListToolName = "task_list"

// TaskListTool is the single mutating tool over the persisted task list.
type TaskListTool struct{}

// NewTaskListTool constructs a TaskListTool.
func NewTaskListTool() *TaskListTool { return &TaskListTool{} }

func (*TaskListTool) Name() string                { return TaskListToolName }
func (*TaskListTool) IsSubagentTool() bool        { return false }
func (*TaskListTool) MayProduceLargeOutput() bool { return false }

func (*TaskListTool) Schema() sporecore.RegistryToolSchema {
	// Fields kept sorted/stable for cache stability: `action` (required) plus the
	// union of per-action fields, listed alphabetically.
	return sporecore.RegistryToolSchema{
		Name:        TaskListToolName,
		Description: "Manage the persisted task list: add, update, complete, or list tasks",
		Parameters: json.RawMessage(`{
			"type": "object",
			"properties": {
				"action": {
					"type": "string",
					"enum": ["add_task", "complete_task", "list_tasks", "update_task"]
				},
				"blockers": {"type": "array", "items": {"type": "integer"}},
				"description": {"type": "string"},
				"id": {"type": "integer"},
				"status": {
					"type": "string",
					"enum": ["blocked", "completed", "in_progress", "pending"]
				}
			},
			"required": ["action"]
		}`),
		// Intentionally NOT ReadOnly: this tool mutates shared on-disk state and
		// must dispatch sequentially. See package docs.
		Annotations: sporecore.ToolAnnotations{},
	}
}

func (t *TaskListTool) Execute(ctx context.Context, call sporecore.ToolCall, _ sporecore.SandboxProvider, toolCtx *sporecore.ToolContext) sporecore.ToolOutput {
	// 1. Parse params (bad input → recoverable).
	var params TaskListParams
	if e := parseParams(call, &params); e != nil {
		return e.ToToolOutput()
	}
	if e := validateTaskListParams(&params); e != nil {
		return e.ToToolOutput()
	}

	// 2. Load current list from the run store, keyed by the STABLE project
	//    namespace (#142) under the shared TaskListExtrasKey (absent → default).
	//    Keying by project_id (not the per-window SessionID the Ralph wrapper
	//    regenerates) is what lets a window reset re-read the prior window's list
	//    instead of re-planning under a session it has never seen. A storage error
	//    or a malformed blob is recoverable.
	list := sporecore.DefaultTaskList()
	value, found, err := toolCtx.GetDurable(ctx, sporecore.TaskListExtrasKey)
	if err != nil {
		return ExecutionFailed(fmt.Sprintf("could not load task list: %s", err), true).ToToolOutput()
	}
	if found {
		if err := json.Unmarshal(value, &list); err != nil {
			return ExecutionFailed(fmt.Sprintf("could not parse task list: %s", err), true).ToToolOutput()
		}
	}

	// 3. Apply the action. Domain errors → recoverable. list_tasks does not
	//    mutate. `added` carries the id assigned by Add so the add branch can
	//    surface it in the success content (#143); it stays nil for the non-add
	//    (and read-only) actions, which return the bare serialized TaskList.
	mutated := false
	var added *uint32
	switch params.Action {
	case TaskListActionAddTask:
		// Capture the assigned id (#143) instead of discarding it. A rejected
		// blocker set still maps to a recoverable error and leaves the list
		// untouched.
		id, domErr := list.Add(*params.Description, params.Blockers)
		if domErr != nil {
			return ExecutionFailed(domErr.Error(), true).ToToolOutput()
		}
		added = &id
		mutated = true
	case TaskListActionUpdateTask:
		if domErr := list.Update(*params.ID, params.Status, params.Description); domErr != nil {
			return ExecutionFailed(domErr.Error(), true).ToToolOutput()
		}
		mutated = true
	case TaskListActionCompleteTask:
		if domErr := list.Complete(*params.ID); domErr != nil {
			return ExecutionFailed(domErr.Error(), true).ToToolOutput()
		}
		mutated = true
	case TaskListActionListTasks:
		// No mutation.
	}

	// 4. Persist the (possibly mutated) list through the run store under the
	//    shared TaskListExtrasKey, keyed by the STABLE project namespace (#142).
	//    list_tasks skips the write.
	if mutated {
		encoded, err := json.Marshal(list)
		if err != nil {
			return ExecutionFailed(fmt.Sprintf("could not serialize task list: %s", err), true).ToToolOutput()
		}
		if err := toolCtx.PutDurable(ctx, sporecore.TaskListExtrasKey, encoded); err != nil {
			return ExecutionFailed(fmt.Sprintf("could not persist task list: %s", err), true).ToToolOutput()
		}
	}

	// 5. Return the serialized current list. On add_task (#143) splice the
	//    assigned id in as a leading `added` key so the success content is
	//    {"added":N,"tasks":[...],"next_id":M} — exactly that field order,
	//    byte-identical across languages. The canonical TaskList marshaling is
	//    {"tasks":[...],"next_id":M} (struct fields marshal in declaration order,
	//    NOT alphabetically), so splicing `"added":N,` right after the opening
	//    brace yields the pinned order deterministically. Re-marshaling a map or
	//    a new struct is NOT used: encoding/json sorts map keys alphabetically
	//    (added,next_id,tasks — wrong order), so we splice onto the existing
	//    canonical bytes. Other actions return the bare TaskList unchanged, and
	//    the PERSISTED blob (step 4) never carries `added`.
	bare, err := json.Marshal(list)
	if err != nil {
		return ExecutionFailed(fmt.Sprintf("could not serialize task list: %s", err), true).ToToolOutput()
	}
	content := string(bare)
	if added != nil {
		// `bare` is the canonical form starting with `{"tasks":`; insert
		// `"added":N,` after the leading `{` (index 1). encoding/json never
		// emits leading whitespace, so byte 0 is always `{`.
		content = fmt.Sprintf("{\"added\":%d,%s", *added, content[1:])
	}
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputSuccess, Content: content}
}

// validateTaskListParams enforces the per-action field requirements that the
// internally-tagged Rust enum gets for free from serde. An unknown action or a
// missing required field is an InvalidParameters error.
func validateTaskListParams(p *TaskListParams) *ToolExecutionError {
	switch p.Action {
	case TaskListActionAddTask:
		if p.Description == nil {
			return InvalidParameters("add_task requires `description`")
		}
	case TaskListActionUpdateTask:
		if p.ID == nil {
			return InvalidParameters("update_task requires `id`")
		}
	case TaskListActionCompleteTask:
		if p.ID == nil {
			return InvalidParameters("complete_task requires `id`")
		}
	case TaskListActionListTasks:
		// No required fields.
	case "":
		return InvalidParameters("missing required field `action`")
	default:
		return InvalidParameters(fmt.Sprintf("unknown action %q", p.Action))
	}
	return nil
}

// Compile-time interface check.
var _ sporecore.Tool = (*TaskListTool)(nil)
